### PR TITLE
libobs + UI: Add Missing Files API & Dialog

### PR DESCRIPTION
**Pull request by VodBox**
### Description

This change makes it so when you open the program or switch to a scene collection, sources will report whether they have any missing files. If they have missing files, a dialog will show to aid the user in finding/replacing/clearing the file paths, and will optionally perform searches in the directories of swapped files to speed up the process.

### Motivation and Context

Whether moving between PCs, accidentally deleting or moving a file that you didn't know was still being used, or importing a collection from another operating system/program, there can be many cases where files end up going missing that affect a persons setup, and may not be noticeable until they're well into a recording/streaming session, or at a point where they can no longer recover the lost file. This brings attention to those issues when the user opens the program, and helps speed up the process of finding files by performing searches in directories where/when other files have been found.

### How Has This Been Tested?

This has been tested on Windows 10 with specially created collections to test all source types, and with imported collections from #1956. I additionally used Azure Pipelines on my forked copy to confirm that the source built on all platforms.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

**Comment by VodBox** [comment]
Windows and macOS builds for this PR can be found under Artifacts [here.](https://dev.azure.com/obsjim/obsjim/_build/results?buildId=2178&view=results)

File: plugins/text-freetype2/text-freetype2.c [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿/******************************************************************************
+/******************************************************************************
```
Comment (jp9000) [inline plugins/text-freetype2/text-freetype2.c]: I assume your editor added a BOM.  We don't like BOMs around these parts.

Comment (VodBox) [reply plugins/text-freetype2/text-freetype2.c]: My editor actually removed a BOM in this instance.
`
-<EF><BB><BF>/******************************************************************************
+/******************************************************************************
`

Comment (jp9000) [reply plugins/text-freetype2/text-freetype2.c]: Ah, whoops.  Not a biggie then.

Comment (mufunyo) [reply plugins/text-freetype2/text-freetype2.c]: But what if someone tried to open the file on a PowerPC?!

**Comment by jp9000** [comment]
This is really nice.  Briefly looking at it, I do wonder however if the dialog feels a bit complicated.  This would work for most users, but I'm a little concerned about intimidating some users.  Also the dialog box could be a bit bigger.

What does everyone else think about simplifying the replacement process?  For example, just making it so they just select the replacement file for each and leave it, and then there's like no apply button and such.  You select the file, it automatically removes it from the list, and then when the user is done they just hit close.  Just sort of streamlining the process or something.

**Comment by dodgepong** [comment]
I don't think removing entries from the list is a good idea, because we want users to have confirmation that the right file is selected. Also, selecting a file in a directory, by default, automatically searches that directory for other files with matching filenames to missing files. I've been trying to discuss ways to make the icons less confusing or even remove the necessity for them, because I can't shake the feeling that they end up making the UI more confusing, not less.

**Comment by jp9000** [comment]
I think it's okay as-is, the only thing I'm concerned about is the experience of this UI.  We need to think about how someone who's not great with computers would feel if they see this.  Every little detail in a user interface can count towards or against that.  I just don't know what those icons mean when I look at them.  It just feels a bit complicated.

Does anyone have any ideas about this?

**Comment by WizardCM** [comment]
Tested this PR in my migration from Windows -> Ubuntu, when loading my imported Scene Collection for the first time. It identified the missing files correctly, and after I selected one file, it correctly fetched the others (which I conveniently moved into a single directory). Overall, success!

As I mentioned in #1956, however, as scripting doesn't provide a UI to remap a path, we may need to add that capability first (feels weird not being able to fix it), and then expose it via this dialog too.

**Comment by jp9000** [comment]
So, how do we make this more user friendly?

**Comment by WizardCM** [comment]
It's a tricky question.

* We could make the Missing File column only show the parent folder and the filename rather than the full path, which'd hopefully make the UI feel less noisy.
* Opening it wider by default would help make paths more readable - I'd recommend no wider than 80% of the display's width

Past that, not a lot we can do. Unless you want to make it a wizard with a step-by-step process, which isn't common for finding missing files, but would improve the experience for users.

**Comment by Heufneutje** [comment]
I'm noticing the same forward slash / backslash mixup in the file paths that also happened in the remux dialog is happening

**Comment by saadbruno** [comment]
I don't think UI at this point is a problem. With file-linking dialogs, I think it is better to show the user as much info as possible, so the user can know exactly which items are missing and why. It might be intimidating at first, but a careful read can already solve this problem.

You have a clear message saying "ops, files are missing!" and a button at the bottom to search directories. That's more than enough for the user to figure out what to do.

Here are 2 other examples of similar dialogs from other softwares:  
## Cinema 4D:
c4d

## Premiere Pro:
premiere

They both show full file paths to the files, when they're in a different directory than the project, and only a relative path when they're in the same directory.

I work with professional production using OBS, and with several producers, and having a dialog showing every missing assets would make our job a lot easier

**Comment by WizardCM** [comment]
Heads up: there are merge conflicts.

**Comment by jp9000** [comment]
This is tight.

**Comment by exeldro** [comment]
The merge is missing documentation changes for struct obs_source_info in https://github.com/obsproject/obs-studio/blob/master/docs/sphinx/reference-sources.rst

**Comment by jp9000** [comment]
Thanks, we'll update it at some point.

**Comment by SplatRT** [comment]
Someone on crowdin has suggested changing

> `MissingFiles.HelpText="You have some missing files since you last used OBS."`

to

> `MissingFiles.HelpText="Some files are missing since you last used OBS."`

**Comment by jp9000** [comment]
Ok, corrected in 219a0913c

**Pull request merged**